KAFKA-13003: In kraft mode also advertise configured advertised port instead of socket port

ueisele opened the pull request:
In Kraft mode Apache Kafka 2.8.0 does advertise the socket port instead of the configured advertised port.

A broker given with the following configuration
```
listeners=PUBLIC://0.0.0.0:19092,REPLICATION://0.0.0.0:9091
advertised.listeners=PUBLIC://envoy-kafka-broker:9091,REPLICATION://kafka-broker1:9091
```
advertises on the _PUBLIC_ listener _envoy-kafka-broker:19092_, however I would expect that _envoy-kafka-broker:9091_ is advertised. In ZooKeeper mode it works as expected.

The reason is that in the BrokerServer at the moment the socket server port is used for registration at the controller: https://github.com/apache/kafka/blob/2beaf9a720330615bc5474ec079f8b4b105eff91/core/src/main/scala/kafka/server/BrokerServer.scala#L286

In KafkaServer class which is used in ZooKeeper mode the configured advertised port is used: https://github.com/apache/kafka/blob/2beaf9a720330615bc5474ec079f8b4b105eff91/core/src/main/scala/kafka/server/KafkaServer.scala#L462

I changed the BrokerServer class, so that in Kraft mode like in ZooKeeper mode also the configured advertised port is registered.

I manually tested it with a Docker-Compose setup. It basically runs 3 Kafka Broker with Apache Kafka 2.8 in Kraft mode and an Envoy proxy in front of them. With Apache Kafka 2.8.0 it does not work, because Kafka does not advertise the configured advertised port. For more details about the setup see: https://github.com/ueisele/kafka/tree/fix/kraft-advertisedlisteners-build/proxy-examples/proxyl4-kafkakraft-bug-2.8

The same Docker-Compose setup with the fix (proposed in this pull request) works and advertises the configured advertised port. For more details see: https://github.com/ueisele/kafka/tree/fix/kraft-advertisedlisteners-build/proxy-examples/proxyl4-kafkakraft-fix-2.8

At the moment there is no dedicated test for BrokerServer class. Therefore I did not create a test so far. Where such a test should be added? Is https://github.com/apache/kafka/blob/trunk/core/src/test/scala/integration/kafka/server/RaftClusterTest.scala a good place?

ueisele committed:
core: in kraft mode also advertise advertised port instead of socket server port

ijuma commented on core/src/main/scala/kafka/server/BrokerServer.scala:
```diff
@@ -283,7 +283,7 @@ class BrokerServer(
           setHost(if (Utils.isBlank(ep.host)) InetAddress.getLocalHost.getCanonicalHostName else ep.host).
           setName(ep.listenerName.value()).
-          setPort(socketServer.boundPort(ep.listenerName)).
+          setPort(ep.port).
```
Thanks for the PR. Can you please add a test?

ueisele replied:
Yes, sure. I will add a test.

ueisele replied:
I added two tests to RaftClusterTest class. One for verifying that the configured advertised host and port are registered if the port is non zero. And another which in verifies that if the advertised port is 0, the bound port is used as actual advertised port. This is required to support binding the listener to any free port, which is used for example for the integration tests.

ueisele committed:
KAFKA-13003: in kraft mode advertise configured advertised port if not zero

ueisele committed:
KAFKA-13003: fix Scala 2.12 incompatibility in RaftClusterTest

ueisele committed:
KAFKA-13003: fix timing issue in RaftClusterTest tests for advertised listeners

It is possible DescribeClusterRequest is processed before the broker is
aware of all other brokers. Therefore we must repeat until the
DescribeClusterResponse contains the expected number of brokers.

ijuma commented:
@ueisele quick heads up that most committers are pretty busy this week due to the 3.0 feature freeze. We should be able to review and merge this next week.

hachikuji commented on core/src/test/scala/integration/kafka/server/RaftClusterTest.scala:
```diff
@@ -313,4 +318,94 @@ class RaftClusterTest {
+      val (runningBrokerServer, foundRunningBroker) = TestUtils.computeUntilTrue(anyBrokerServer(cluster)){
+        brokerServer => brokerServer.currentState() == BrokerState.RUNNING
+      }
+      assertTrue(foundRunningBroker, "No Broker never made it to RUNNING state.")
```
nit: never -> ever?

hachikuji commented on core/src/test/scala/integration/kafka/server/RaftClusterTest.scala:
```diff
@@ -313,4 +318,94 @@ class RaftClusterTest {
+      assertTrue(foundRunningBroker, "No Broker never made it to RUNNING state.")
+
+      val (describeClusterResponse, metadataUpToDate) = TestUtils.computeUntilTrue(
+        sendDescribeClusterRequest(runningBrokerServer.socketServer, nodes.externalListenerName())
```
Maybe worth a comment that `connectAndReceive` connects to the bound port instead of the advertised port since that is not obvious.

hachikuji commented on core/src/test/scala/integration/kafka/server/RaftClusterTest.scala:
```diff
@@ -313,4 +318,94 @@ class RaftClusterTest {
+      ) {
+        response => response.nodes().size() == nodes.brokerNodes().size()
+      }
+      assertTrue(metadataUpToDate, s"Broker never reached expected cluster size of ${nodes.brokerNodes().size()}")
```
nit: these tests are basically the same except for the initialization and assertions. It would be nice to factor out a helper or two.

ueisele replied:
I refactored the two tests and also considered your other two remarks.

hachikuji reviewed (commented):
@ueisele Thanks for the fix! Looks good overall, but left some small comments.

ueisele committed:
KAFKA-13003: Refactor tests for advertised listener in RaftClusterTest

ueisele commented:
I checked the tests which are failing. For me it looks like they are not related to the change made in this pull request.
What is failing is "RaftClusterTest.testCreateClusterAndCreateAndManyTopicsWithManyPartitions" for which there is already a ticket (https://issues.apache.org/jira/browse/KAFKA-12629) and "RaftEventSimulationTest". I do not know why the latter one fails, but it does not use the BrokerServer.

hachikuji reviewed (approved):
LGTM. Thanks for the patch!
